### PR TITLE
#29996 #29998 Improved path cache syncing

### PR DESCRIPTION
**Pull request by manneohrstrom**
Handles edge cases and fixes bugs around path cache syncing.

- FilesystemLocation records in Shotgun whose paths do not fit inside any of the storages defined in the `roots.yml` are now ignored in the sync (a log message is generated). This is an edge case that you'll hit if your roots.yml file is not the same as your local storage def in Shotgun (a setup which we strongly advise against having).
- Fixed a bug causing a full path cache sync to kick in every single time folder creation was running. It turns out that when running on multiple machines, with different people creating folders, the cache sync would always fall back onto a full sync. Fixing this bug will mean a big improvement in app launch speed and users will stop seeing the "hang on, toolkit is generating folders"
- Changed the behaviour so that `Toolkit_Folder_Create` event log records are only created when items were actually added to the `FilesystemLocation` table in Shotgun. Previously, it would always add a record (with an empty list of folder references), resulting in unnecessary syncing and event log records.
- Added unit tests for all cases above and improved some of the other tests slightly.
- Added better logging to the sync process.

### CODE REVIEW

File: python/tank/path_cache.py [hunk 1]
```diff
@@ -344,15 +344,22 @@ def synchronize(self, log=None, full_sync=False):
...
+            # by this query should be the last one previously processed by the 
+            # path cache (via the event_log_id variable)
             response = self._tk.shotgun.find("EventLogEntry", 
                                              [ ["event_type", "in", ["Toolkit_Folders_Create", 
                                                                      "Toolkit_Folders_Delete"]], 
                                                ["id", "greater_than", (event_log_id - 1)],
                                                ["project", "is", self._get_project_link()] ],
                                              ["id", "meta", "event_type"],
-                                             [{"field_name": "id", "direction": "desc"}] )   
+                                             [{"field_name": "id", "direction": "asc"}] )
```
Comment (manneohrstrom) [inline python/tank/path_cache.py]: This is the big, critical change that will cause the sync to be incremental rather than full!

File: python/tank/path_cache.py [hunk 2]
```diff
@@ -586,29 +600,28 @@ def _do_incremental_sync(self, cursor, log, sg_data):
...
                 # this is a creation request! Replay it on our database
                 created_folder_ids.extend( d["meta"]["sg_folder_ids"] )
             else:
                 # should never come here
                 raise Exception("Unsupported event type '%s'" % d)
-                
-        if len(created_folder_ids) == 0:
-            # one or more folder creation events were detected but none of them had actually
-            # resulted in any actual folders being created!
-            return []
```
Comment (manneohrstrom) [inline python/tank/path_cache.py]: The reason this early exit is omitted is so that the code can go in and complete the sync and add the sqlite path cache event log tracking marker to say "i have synced my records up to this point"

File: python/tank/path_cache.py [hunk 3]
```diff
@@ -644,9 +662,17 @@ def _replay_folder_entities(self, cursor, log, max_event_log_id, ids=None):
                                    SG_ENTITY_TYPE_FIELD, 
                                    SG_ENTITY_NAME_FIELD],
                                   [{"field_name": "id", "direction": "asc"},])
+        elif ids == []:
+            # incremental sync but with no folders
+            self._log_debug(log, "No folders need to be replayed, won't fetch anything from Shotgun...")
```
Comment (manneohrstrom) [inline python/tank/path_cache.py]: performance: one less `find` call in a common scenario.

**Comment by manneohrstrom** [comment]
Added some comments to help explain some of the changes

File: python/tank/path_cache.py [hunk 4]
```diff
@@ -365,14 +372,21 @@ def synchronize(self, log=None, full_sync=False):
                 if r["event_type"] == "Toolkit_Folders_Delete":
                     num_deletions += 1
                     
-            if len(response) == 0 or response[0]["id"] != event_log_id:
+            self._log_debug(log, "Event log contains %s creations and %s deletions" % (num_creations, num_deletions))
+            
+            first_received_event_log_id = response[0]["id"]
```
Comment (alandann) [inline python/tank/path_cache.py]: this'll fail if response is an empty list

File: python/tank/path_cache.py [hunk 5]
```diff
@@ -644,9 +662,17 @@ def _replay_folder_entities(self, cursor, log, max_event_log_id, ids=None):
                                    SG_ENTITY_TYPE_FIELD, 
                                    SG_ENTITY_NAME_FIELD],
                                   [{"field_name": "id", "direction": "asc"},])
+        elif ids == []:
```
Comment (alandann) [inline python/tank/path_cache.py]: interesting style choice ;) - would have probably written it like this:

elif len(ids) == 0:

File: python/tank/path_cache.py [hunk 6]
```diff
@@ -644,9 +662,17 @@ def _replay_folder_entities(self, cursor, log, max_event_log_id, ids=None):
...
+            # incremental sync but with no folders
+            self._log_debug(log, "No folders need to be replayed, won't fetch anything from Shotgun...")
+            sg_data = []
```
Comment (alandann) [inline python/tank/path_cache.py]: can I suggest defining sg_data outside the if-else block - it would be more robust should the code change at a later date.

File: python/tank/path_cache.py [hunk 7]
```diff
@@ -749,13 +791,15 @@ def _replay_folder_entities(self, cursor, log, max_event_log_id, ids=None):
             else:
                 # Note: edge case - for some reason there was already an entry in the path cache
                 # representing this. This could be because of duplicate entries and is
-                # not necessarily an anomaly.
+                # not necessarily an anomaly. It could also happen because a previos sync failed
+                # at some point half way through.
                 self._log_debug(log, "Found existing record for '%s', %s. Skipping." % (local_os_path, entity))
                 pass
```
Comment (alandann) [inline python/tank/path_cache.py]: excruciatingly minor but the pass is now redundant!

File: python/tank/path_cache.py (hunk 2 above)
Comment (alandann) [inline python/tank/path_cache.py]: Might be worth adding a comment to that effect in the code so that someone doesn't come along at a later date and optimise it!

**Comment by alandann** [comment]
One potential bug and a few minor suggestions but otherwise looks good.

File: python/tank/path_cache.py (hunk 4 above)
Comment (manneohrstrom) [inline python/tank/path_cache.py]: nice find!

File: python/tank/path_cache.py (hunk 6 above)
Comment (manneohrstrom) [inline python/tank/path_cache.py]: done!

File: python/tank/path_cache.py (hunk 4 above)
Comment (manneohrstrom) [inline python/tank/path_cache.py]: fixed and added another unit test for this edge case.

File: python/tank/path_cache.py (hunk 7 above)
Comment (manneohrstrom) [inline python/tank/path_cache.py]: sorry, completely missed that!
